### PR TITLE
Lets the user specify us dropping hints in response, resolves #1789.

### LINKED ISSUE #1789: Add option to disable hint_data output.
When using the Match service, the response contains hint_data. This additional data doubles the size of the response from the match query and is not needed unless you're also using the viaroute service. It would be nice to have a command line option or service request parameter that disables the output of hint_data at least for the Match service.

### PR DESCRIPTION
**Pull request by daniel-j-h**
For #1789.

Adds an `?emit_hints=false` option which lets us skip generating and
emitting hints for Waypoints. This can be used to decrease the response
size when the user does not need hints anyway.

We should think about making `false` the default in v6, and let advanced
users be explicit about the hint feature.

## Tasklist
 - [x] figure out a good name for this option, then document in http.md api spec
 - [x] review
 - [x] adjust for comments

cc @TheMarex

### CODE REVIEW

**Comment by TheMarex** [comment]
>  figure out a good name for this option, then document in http.md api spec

`generate_hints` maybe?

> We should think about making false the default in v6, and let advanced
users be explicit about the hint feature.

:+1: Can you ticket this and put it in the v6 API changes ticket?

**Comment by emiltin** [comment]
what about simply hints=false?

**Comment by daniel-j-h** [comment]
`hints` is already taken - it's the place where users can send hints previously received.

Check the shiny new docs for the `hints` parameter:
http://project-osrm.org/docs/v5.5.0/api/#general-options

File: include/server/api/base_parameters_grammar.hpp [hunk 1]
```diff
@@ -141,11 +141,18 @@ struct BaseParametersGrammar : boost::spirit::qi::grammar<Iterator, Signature>
...
+        emit_hints_rule =
+            qi::lit("emit_hints=") >
+            qi::bool_[ph::bind(&engine::api::BaseParameters::emit_hints, qi::_r1) = qi::_1];
+
         bearings_rule =
             qi::lit("bearings=") >
             (-(qi::short_ > ',' > qi::short_))[ph::bind(add_bearing, qi::_r1, qi::_1)] % ';';
 
-        base_rule = radiuses_rule(qi::_r1) | hints_rule(qi::_r1) | bearings_rule(qi::_r1);
+        base_rule = radiuses_rule(qi::_r1)   //
```
Comment (oxidase) [inline include/server/api/base_parameters_grammar.hpp]: I have recently found `// clang-format off` `// clang-format on` and used here https://github.com/Project-OSRM/osrm-backend/blob/opening_hours/include/util/opening_hours.hpp#L302

Would it be ok to allow suppress clang format for qi rules  and use qi-usual formating like
```
    emit_hints_rule
        = qi::lit("emit_hints=")
        > qi::bool_[ph::bind(&engine::api::BaseParameters::emit_hints, qi::_r1) = qi::_1]
        ;
```
?

Comment (daniel-j-h) [inline include/server/api/base_parameters_grammar.hpp]: Ah yeah makes sense probably not only for Spirit grammars but e.g. also for program option parser and other similar code. Want to ticket this or make a separate pull request for it?

**Comment by daniel-j-h** [comment]
Adjusted to comments - @oxidase and @themarex could you have a final look again.

**Comment by TheMarex** [comment]
Flagging here: This will need an adaption in `node-osrm`.

**Comment by TheMarex** [comment]
There is one test still failing https://travis-ci.org/Project-OSRM/osrm-backend/jobs/184548957#L5796

**Comment by daniel-j-h** [comment]
Ah sorry, missed updating the expected parsing failure for the `emit_hints` -> `generate_hints` name change. Updating and then merging this.
